### PR TITLE
Bonsai 2 27B: lossless performance and API stability qualification

### PR DESCRIPTION
**Pull request by jonathan308**
## Summary

**[Swift #28](https://github.com/Layr-Labs/mlx-swift/pull/28)** and
**[SDK #155](https://github.com/Layr-Labs/mlx-swift-lm/pull/155)** are merged.
This provider now pins their actual merged commits. This is a follow-up to the
merged support PR, with the qualified default-on performance profile retained.

Performance and local API-stability follow-up to merged Bonsai support #1124.
Compose exact public Swift/SDK revisions, add native media/cache qualification
fixtures and repair local typed-error/nested-reasoning/late-stream handling.
No model publication, catalog, signing/release, coordinator or deployment change.

## Default-on update

Both eligible optimizations now enable when their overrides are **unset** or
exactly `1`. Explicit `0` restores the original path; other explicit spellings
remain disabled. Diagnostics stay opt-in and the generic cache kill switch remains.

Fresh M3 optimized component tests: **26/26 pass with overrides unset**, **26/26
with explicit1**; OFF and invalid profiles each pass22 with four deliberately
skipped ON-only scheduling tests. Real packed-layer outputs, parameter/stream/
tracing invariants, eligibility, deferred fills, write faults and retirement
are covered. Weights, dtypes, math, scope gates and reported tradeoffs are unchanged.

## Linked issue

Follow-up to #1124; no separate issue or automatic closure.

## Before

```mermaid
flowchart LR
  Provider[Merged support] --> SDK[Lazy compact carry]
  SDK --> Swift[Repeated constant widening]
  Swift --> Output[Native Bonsai output]
  Invalid[Negative local budget] --> Wrong[Outside typed SDK error mapping]
  Thought[Nested reasoning] --> Flat[Flat routing policy]
  Late[Error after headers] --> Cut[Incomplete HTTP body]
```

## After

```mermaid
flowchart LR
  Provider[Public immutable dependency pins] --> Profile[Qualified default performance profile]
  Profile --> SDK[Earlier exact carry submission]
  Profile --> Swift[Exact constant reuse]
  SDK --> Output[Same native arithmetic and output]
  Swift --> Output
  Invalid[Negative budget] --> Reject[Typed HTTP400 before inference]
  Thought[Qualified Bonsai reasoning] --> Policy[Existing bounded nested policy]
  Policy --> Opaque[Preserve reasoning and opaque tool arguments]
  Late[Error after headers] --> Frame[Sanitized terminal SSE error]
  Frame --> End[Complete transfer; no fabricated success]
```

## Full benchmarks

[Full report and sanitized measurement summaries](https://github.com/jonathan308/d-inference-nemotron/blob/e0677f6f19fecbef07daca68b9995323c1a7ca6d/docs/reports/2026-09-18-bonsai2-lossless-performance.md).

Native paged B1, prefix tiers OFF, repeated ABBA, unchanged weights, fixed128
timed decode tokens plus first token. Rates are medians in tok/s; prefill here
means prompt / submit-to-first-token time, **not pure kernel timing**.

| Host | Prompt tokens | Effective prefill OFF → ON | Decode OFF → ON |
|---|---:|---:|---:|
| M3 Ultra | 1,024 | 292.67 → 293.93 | 29.49 → 43.32 |
| M3 Ultra | 10,000 | 280.58 → 282.33 | 28.91 → 39.80 |
| M3 Ultra | 20,000 | 265.13 → 267.06 | 28.77 → 36.00 |
| M3 Ultra | 50,000 | 222.52 → 225.08 | 23.89 → 28.10 |
| M5 Max | 1,024 | 654.96 → 669.92 | 38.24 → 46.33 |
| M5 Max | 10,000 | 500.12 → 506.87 | 31.07 → 37.91 |
| M5 Max | 20,000 | 425.35 → 413.12 | 24.77 → 28.40 |
| M5 Max | 50,000 | 338.45 → 330.89 | 17.43 → 18.99 |

All129 token IDs match. Actual authenticated synthetic tool-history API decode
improves49–52% on M3 and23–25% on M5. Full API prefill/decode, native memory
tables and latest stable M3 20K confirmation are in the report.

## Test plan

- [x] Frozen numerical/state oracles, native cohort/concurrency/cancellation,
  cache fixtures and ownership/reload, with exact artifact/binary scopes recorded.
- [x] Swift/SDK suites and updated server213/provider-router15 cases.
- [x] Ordinary serving15/15, lifecycle3/3, negative error framing2/2.
- [x] Original M3 encrypted checkpoint fixture and6/6 isolated filters after
  real disk cleanup, preserving the safety reserve and original assertions.
- [x] Initial public-promotion source equality; default follow-up changes only enablement policy, tests/docs and pins.
- [x] make docs-impact-check BASE=public/master; make docs-check (307 documents);
  signed commit, public dependency reachability and diff whitespace.
- [x] Pre-default-activation public-pin release build:287.54s; executable SHA256 identical
  to the final qualified runtime, not a replacement benchmark or quality claim.
- [ ] Strict tool-success matrix fully green:24/28; four baseline copy failures remain.
- [ ] Full provider suite all-green:four baseline GPTOSS cases/129 numerical issues.
- [ ] Release-signed persistent cache and hosted/trust qualification.
- [ ] All required end-to-end release gates complete, with current evidence.

## Components touched

- [x] provider-swift
- [x] docs and dependency pins
- [ ] coordinator
- [ ] console-ui
- [ ] enclave
- [ ] infra / CI / release

## Protocol / interface changes

Local HTTP negative-budget validation and late streaming errors now preserve
typed/sanitized failure semantics. No coordinator wire or catalog changes,
no version bump, no data migration. The SDK's original direct service signature
and cancellation behavior remain intact.

## Documentation impact

- [x] Canonical build/test/configuration/CLI/direct-mode documentation, changelog
  and linked benchmark report updated together.

## Composition and notes for reviewers

- Provider: `ab44396397429d8f7029f390837bda24444f82c1`. Reconciled with
  current master `1451a4c8911bc4ec9d094c31fa508f22e55d1865` (0.9.7).
  Only the changelog, test-document placement and CLI freshness stamp conflicted;
  both sides' intended content is retained, and the stale CLI version is corrected.
  The four Bonsai feature source files and existing test bodies are unchanged.
  The incoming master changes are preserved.
- SDK: `e22fc82bdb7bfbd93874d56c7df9ca3306782b09`.
- Swift: `0f4fe403bef6899e8a72882bc6d4036a7a62ae31`.
- Both dependency trees are exactly identical to the previously qualified public
  review revisions (`b16fc923` SDK and `d7c1d3dc` Swift). MLX core/C pins,
  default activation and model weights are unchanged. Dependencies use public
  Layr-Labs URLs and immutable merged commits.
- Merge-resolution checks: release-version parity at 0.9.7, diff checks,
  documentation impact (21 changed files), and 310-document lint pass.
  Source-equivalent repinning does not relabel old benchmarks as new runs;
  exact-head public CI remains a separate check.
- Both switches now default ON for eligible paths; explicit0 disables. Retained converted constants cost about1.60GB.
  Lower active-memory peak is not always lower process peak, especially at50K.
- Rollback performance switches to0 for original paths; API fixes need their
  separately reviewed rollback. No weights or persistence format changed.

## Qualification scope and limitations

[Full two-host benchmark and regression report](https://github.com/jonathan308/d-inference-nemotron/blob/e0677f6f19fecbef07daca68b9995323c1a7ca6d/docs/reports/2026-09-18-bonsai2-lossless-performance.md) includes sanitized
measurement summaries, exact artifact identities, all 1K/10K/20K/50K prefill,
decode and memory results, and authenticated synthetic tool-history results.

- Exact frozen logits/recurrent state/paged-KV comparisons pass on M3 Ultra and
  M5 Max, B1/B2. Native multimodal/cohort, concurrency, cancellation, readmission,
  prefix-cache fixtures and ownership/reload checks pass within recorded scope.
- Ordinary serving15/15, lifecycle3/3 and negative late-error framing2/2 pass.
  Original strict tool-success matrix remains **24/28**, with four copy-value
  stress failures reproduced with optimizations OFF/frozen controls.
- Full provider suite retains four GPTOSS cases/129 numerical issues reproduced
  on the independent frozen baseline. Signed-persistent/hosted/trust gates remain open.
- Both performance switches are **on by default for eligible paths**. No weights, precision,
  model architecture, MTP capability, context limit or numerical tolerance changes.
- Large measured decode gains do not imply a large prefill gain: M3 prefill is
  roughly +0.4–1.2%; longer M5 cells decrease about2–3%. Whole-process peak can
  increase even when active allocation decreases.
- [ ] All required end-to-end release gates complete, with current evidence.

The initial public-pin promotion rebuilt source-equivalently in287.54s to
qualified executable `0db475ecee9f0d5cc96d37c8892c871641cdd1dd79eb301218b2fdb8f2f37cfc`.
That receipt predates this default-activation follow-up and is not relabelled as
a newly built default-on executable. The follow-up changes enablement policy,
tests, docs and pins, not numerical kernels or eligibility. Fresh M3 component
tests qualify unset/default, explicit-ON, explicit-OFF and invalid-value profiles;
no full model server was restarted or benchmark rerun for this narrow update.

GitHub checks are still progressing. Vercel preview checks require external
deployment authorization; no deployment approval or rules bypass was performed.

The documentation-only upstream conflict is resolved. The ordinary pre-push
Go tests, TypeScript lint and Next.js build pass with repository-selected Go1.25
and Node22; no hook was bypassed. The earlier system-Go1.27 JSON-encoding failures
were preserved and reproduced, then passed unchanged assertions on the pinned
toolchain. No coordinator/runtime fix was made to force those results.

### CODE REVIEW

**Comment by brandon-eigenlabs** [comment]
Independent M4 Max corroboration:

- Apple M4 Max, 40 GPU cores, 64 GB
- Model SHA: `ea1e901e4946c0ba9ad70c78517548808b353db6b3a13e87a8fa20468d81244c`
- Provider: `8d7142dd0ccdaab6e2080edefd413196b067f5bb`
- Process-level ABBA, paged B1, 20K prompt, prefix cache off, 128 timed decode tokens

| Metric | OFF | ON | Change |
|---|---:|---:|---:|
| Decode | 21.25 tok/s | 25.40 tok/s | +19.53% |
| Effective prefill | 152.39 tok/s | 151.14 tok/s | −0.82% |
| MLX peak | 19.18 GB | 17.01 GB | −11.29% |

All four runs produced identical 129-token arrays (`32169ac627b2b2307975a540622b69cfb165c0665fbb93c198e292effebcf93d`). This independently corroborates the optimization's performance direction and output stability on M4 Max. It does not requalify current head `ec262c10` or its newer dependency pins/default-on policy.

**Review by Gajesh2007 — APPROVED** [review]
